scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -192,6 +193,8 @@ sealed trait Execution[+T] extends java.io.Serializable {
    */
   def zip[U](that: Execution[U]): Execution[(T, U)] =
     Zipped(this, that)
+
+  override val hashCode = super.hashCode
```

Comment:
You're correct that this currently defers to `Object`, so it breaks caching.  In theory we could build custom `hashCode`s on each subclass and make it a val, but I don't feel great about that approach.